### PR TITLE
replace util.UniqRands

### PR DESCRIPTION
**Pull request by vearne**
The performance of util.UniqRands is not good, I re-implemented

### CODE REVIEW

**Comment by vearne** [comment]
|action|Previous implementation (ns/op)|my implementation (ns/op)|
|:---|:---|:---|
|select 20 from 100|2793|788|
|select 50 from 100|7812|1729|
|select 20 from 1000|2621|2152|
|select 50 from 1000|2621|3030|

**Comment by vearne** [comment]
|action|Previous implementation (ns/op)|my implementation (ns/op)|
|:---|:---|:---|
|select 20 from 100|2793|788|
|select 50 from 100|7812|1729|
|select 20 from 1000|2621|2152|
|select 50 from 1000|6764|3030|

File: internal/util/rand.go [hunk 1]
```diff
@@ -4,19 +4,23 @@ import (
...
-	for {
-		num := rand.Intn(n)
-		if _, ok := set[num]; !ok {
-			set[num] = struct{}{}
-			nums = append(nums, num)
-		}
-		if len(nums) == l {
-			goto exit
-		}
+func UniqRands(count int, choiceCount int) []int {
```
Comment (ploxiln) [inline internal/util/rand.go]: This switches the order of the arguments, but the caller still uses the original argument order.

I see you don't like `l` and `n` and that's fair, they're not super clear. But I find "count" being in both of your argument names to be confusing. I think we can come up with better names. Here's a proposal (back in the original order): `(quantity, maxval int)`

**Comment by ploxiln** [comment]
Since the default QueueScanSelectionCount is 20, I think the interesting benchmark cases are:
  * select 20 from 5
  * select 20 from 20
  * select 20 from 50

And you could put the little benchmark function in `internal/util/util_test.go` (see for example `nsqd/guid_test.go`)

**Comment by ploxiln** [comment]
By the way, you can push updates to your branch, and the updates will show up in this pull request, so you don't have to close it and open a new one.

File: internal/util/rand.go (hunk 1 above)
Comment (vearne) [reply internal/util/rand.go]: I accept your opinion, and I will fix it.

**Comment by vearne** [comment]
@ploxiln please review

File: internal/util/rand.go [hunk 2]
```diff
@@ -2,21 +2,34 @@ package util
...
-		num := rand.Intn(n)
-		if _, ok := set[num]; !ok {
-			set[num] = struct{}{}
-			nums = append(nums, num)
-		}
-		if len(nums) == l {
-			goto exit
-		}
+func init() {
+	rand.Seed(time.Now().UnixNano())
```
Comment (ploxiln) [inline internal/util/rand.go]: The rand seed is set in `apps/nsqd/nsqd.go`, so we should not do that here.

I think we don't need to set a "good" random seed for tests (and they will be somewhat more deterministic if we don't).

File: internal/util/rand.go [hunk 3]
```diff
@@ -2,21 +2,34 @@ package util
...
+	intSlice := make([]int, maxval)
+	for i := 0; i < maxval; i++ {
+		intSlice[i] = i
+	}
+
+	idx := 0
+	for i := 0; i < quantity; i++ {
+		idx = rand.Int()%maxval + i
+		// swap
+		swap(intSlice, i, idx)
```
Comment (ploxiln) [inline internal/util/rand.go]: I don't think it's worth both a comment and a function for this.
I would just put the line of code that does the swap here.

**Comment by ploxiln** [comment]
Thanks for the updates.

@mreiferson I think this change is worth it because the old implementation could have occasionally (randomly) bad worst-case performance - it could get "unlucky" and not randomly pick the one remaining possibility for a while. So, where it is used, the less-than-optimal average performance is not a problem at all. But the on-rare-occasion-terrible performance is worrysome. This new algorithm will always take a deterministic bounded number of steps (and the average performance is also better).

I had to tweak the benchmark to make it work with the old implementation, because the old implementation did not _internally_ check for `l > n`:

```diff
+++ b/internal/util/util_test.go
@@ -6,7 +6,7 @@ import (
 func BenchmarkUniqRands20_5(b *testing.B) {
        for i := 0; i < b.N; i++ { //use b.N for looping
-               UniqRands(20, 5)
+               UniqRands(5, 5)
        }
@@ -30,7 +30,7 @@ func TestUniqRands(t *testing.T) {
                t.Errorf("error, expect:%v, got:%v\n", 3, len(x))
        }
 
-       x = UniqRands(10, 5)
+       x = UniqRands(5, 5)
        if len(x) == 5 {
                t.Logf("success, %v", x)
```

And the results are similar to those posted above. Before:

```
BenchmarkUniqRands20_5-4    	 2000000	       942 ns/op
BenchmarkUniqRands20_20-4   	  200000	      6889 ns/op
BenchmarkUniqRands20_50-4   	  500000	      3413 ns/op
```

After:

```
BenchmarkUniqRands20_5-4    	 5000000	       237 ns/op
BenchmarkUniqRands20_20-4   	 2000000	       835 ns/op
BenchmarkUniqRands20_50-4   	 2000000	       892 ns/op
```

**Comment by mreiferson** [comment]
LGTM, I’ll leave it for plo to merge when ready.

**Comment by ploxiln** [comment]
I screwed this up, sorry - I was trying to push one another commit to vearne's branch, but:
  * pushed my master branch to his master branch
    * which I was allowed to do because his master branch was the branch used for this open PR in this project of which I am an admin
  * pushing master made his master no longer contain any different commits, and so closed this PR
  * this PR closed, I can no longer push to his (master) branch

so I'll have to do it in another PR, incoming

**Comment by ploxiln** [comment]
replaced with #1019

**Pull request closed without merging**